File: leetcode/1_two_sum/main.go
package main

import "fmt"

func main() {
	fmt.Println(twoSum([]int{2, 7, 9}, 9))
	fmt.Println(twoSum([]int{0, 4, 2}, 6))
	fmt.Println(twoSum([]int{3, 3}, 6))
}

func twoSum(nums []int, target int) []int {
	hashMap := make(map[int]int)

	for i, n := range nums {
		
		if v, ok := hashMap[n]; !ok {
			hashMap[target-n] = i
		} else {
			return []int{v, i}
		}

	}

	return nil
}

File: leetcode/1_two_sum/main.rs
use std::collections::HashMap;
use std::convert::TryInto;


fn main() {
    println!("{:?}", two_sum(vec![2, 7, 9], 9));
    println!("{:?}", two_sum(vec![2, 7, 4], 6));
    println!("{:?}", two_sum(vec![3, 3], 6));
}

fn two_sum(nums: Vec<i32>, target: i32) -> Vec<i32> {
    let mut h_map = HashMap::new();

    for (i, y) in nums.iter().enumerate() {
        let x = target - y;

        if let Some(&j) = h_map.get(&x) {
            return vec![j as i32, i as i32];
        }

        h_map.insert(y, i);
    }
    
    vec![0, 0]
}